### PR TITLE
 [DeadCode] Skip key used in next stmt on RemoveUnusedForeachKeyRector 

### LINKED ISSUE #8255: Ruleset DEAD_CODE original logic violation
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | 0.18.4             |

## Minimal PHP Code Causing Issue

```php
/**
 * Returns last key of any tranversable container.
 */
function lastKey ($container, $defKey = null) {
  $result = $defKey;

  foreach ($container as $result => $value) {
    // next
  }

  return $result;
}
```

## Expected Behaviour

Rector MUST NOT change `foreach ($container as $result => $value) {` into `foreach ($container as $value) {`. We lose `$result = next key` logic from FOR statement. Thus the code always returns default key. Dead code elimination MUST NOT change application logic and it seems to break it in many cases as I see.

### PR DESCRIPTION
**Pull request by samsonasik**
Fixes https://github.com/rectorphp/rector/issues/8255

### CODE REVIEW

**Comment by samsonasik** [comment]
Fixed 🎉 /cc @ethernidee

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I am merging it ;)

File: rules/DeadCode/Rector/Foreach_/RemoveUnusedForeachKeyRector.php [hunk 1]
```diff
@@ -48,31 +51,51 @@ public function getRuleDefinition(): RuleDefinition
...
+            $isNodeUsed = (bool) $this->betterNodeFinder->findFirst(
+                $stmt->stmts,
+                fn (Node $node): bool => $this->nodeComparator->areNodesEqual($node, $keyVar)
+            );
+
+            if ($isNodeUsed) {
+                continue;
+            }
+
+            if ($this->stmtsManipulator->isVariableUsedInNextStmt($node, $key + 1, (string) $this->getName($keyVar))) {
```
Comment (staabm) [inline rules/DeadCode/Rector/Foreach_/RemoveUnusedForeachKeyRector.php:87]: does this mean, the logic only works when the var usage is directly after the loop?

the initial problem also exists on e.g.

```
<?php

namespace Rector\Tests\DeadCode\Rector\Foreach_\RemoveUnusedForeachKeyRector\Fixture;

final class SkipKeyUsedInNexStmt
{
    function lastKey ($container, $defKey = null) {
        $result = $defKey;

        foreach ($container as $result => $value) {
          // next
        }

       $x = 1;
       $y = 2;

        return $result;
      }
}
```

Comment (samsonasik) [reply rules/DeadCode/Rector/Foreach_/RemoveUnusedForeachKeyRector.php:87]: I mean no, that checked with array_slice() :)

Comment (samsonasik) [reply rules/DeadCode/Rector/Foreach_/RemoveUnusedForeachKeyRector.php:87]: The logic is check all next stmts:

```
        $stmts = array_slice($stmtsAware->stmts, $jumpToKey, null, true);
        return (bool) $this->betterNodeFinder->findVariableOfName($stmts, $variableName);
```

**Pull request merged**